src/store/notifications/saga.ts

```diff
@@ -21,13 +31,81 @@ export function* fetch(action) {
   yield put(setStatus(AsyncListStatus.Idle));
 }
 
+export function createEventChannel(userId, pusherClient = new PusherClient()) {
+  return eventChannel((emit) => {
+    const events = pusherClient.events.map((event) => {
+      return {
+        key: event,
+        callback: (notification) => {
+          // ToDo: return user data in notification
+          emit({ ...notification, isUnread: true });
+        },
+      };
+    });
+
+    pusherClient.init(userId, events);
+
+    const unsubscribe = () => {
+      pusherClient.disconnect();
+    };
+
+    return unsubscribe;
+  });
+}
+
+export function* watchForChannelEvent(userId) {
+  const notificationChannel = yield call(createEventChannel, userId);
+
+  while (true) {
+    const { abort, notification } = yield race({
+      abort: take(SagaActionTypes.CancelEventWatch),
+      notification: take(notificationChannel),
+    });
+
+    if (abort) {
+      notificationChannel.close();
+      return false;
+    }
+
+    if (relevantNotificationTypes.includes(notification.notificationType)) {
+      yield call(addNotification, notification);
+    }
+  }
+}
+
+export function* addNotification(notification) {
+  const existingNotifications = yield select(rawNotificationsList);
+
+  yield put(
+    receive([
+      notification,
+      ...existingNotifications,
+    ])
+  );
+}
+
 export function* clearNotifications() {
   yield all([
     put(removeAll({ schema: schema.key })),
     put(receive([])),
   ]);
 }
 
+export function* authWatcher() {
```

Comment:
Something I forgot to test with my stuff....does this work even when you just switch wallet accounts via metamask? I.E., is there a logout/login combination?

Reply:
When we switch addresses we `terminate` before `authorizing`. I did a smoke test with Pusher in debug mode and verified the client is disconnect / reconnecting with the new user.

```
Handling 'accountsChanged' event with payload ['0x4d054f8c925e992518861e0c1931b212aea2320d']
Emitting 'Web3ReactUpdate' with payload {account: '0x4d054f8c925e992518861e0c1931b212aea2320d'}
Pusher :  : ["State changed","connected -> disconnected"]
Pusher :  : ["State changed","initialized -> connecting"]
Pusher :  : ["Connecting",{"transport":"ws","url":"wss://ws-us2.pusher.com:443/app/8370d96dd06fa311e918?protocol=7&client=js&version=8.0.2&flash=false"}]
Pusher :  : ["State changed","connecting -> connected with new socket ID 23773.13929648"]
Pusher :  : ["Event sent",{"event":"pusher:subscribe","data":{"auth":"","channel":"notifications-ce60cf6c-baff-424c-9f5a-058d2ee023c4"}}]
Pusher :  : ["Event recd",{"event":"pusher_internal:subscription_succeeded","channel":"notifications-ce60cf6c-baff-424c-9f5a-058d2ee023c4","data":{}}]
Pusher :  : ["No callbacks on notifications-ce60cf6c-baff-424c-9f5a-058d2ee023c4 for pusher:subscription_succeeded"]
Pusher :  : ["Event sent",{"event":"pusher:subscribe","data":{"auth":"8370d96dd06fa311e918:de5c24adc35d4e5c4bf990d5d19718c1f28e42f6ac780117bfc3118806074575","channel_data":"{\"user_id\":\"ce60cf6c-baff-424c-9f5a-058d2ee023c4\"}","channel":"presence-channel"}}]
Pusher :  : ["Event recd",{"event":"pusher_internal:subscription_succeeded","channel":"presence-channel","data":{"presence":{"count":2,"ids":["7cb0000d-cf0c-4ca7-81ce-8ce72649c4fd","ce60cf6c-baff-424c-9f5a-058d2ee023c4"],"hash":{"7cb0000d-cf0c-4ca7-81ce-8ce72649c4fd":null,"ce60cf6c-baff-424c-9f5a-058d2ee023c4":null}}}}]
```